---
title: "Hw 3, Part 1"
author: "Mrinal Vashisth"
date: "10/3/2019"
output: html_document
---

```{r}
# Part 1
# Imports
library(ggplot2)
library("ape")
library("ggdendro")
library(alignfigR)
library(seqinr)
# BiocManager::install('alignfigR')
# BiocManager::install('ggdendro')
# BiocManager::install('DECIPHER')
library(DECIPHER)
# BiocManager::install('muscle')
library(muscle)
library(annotate)
library(msa)

```

```{r}
#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  # 
## Get annotation
#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  # 

print('Our query is: 
the fasta file')

print('Result:
      
     Lineage (full): cellular organisms; Eukaryota; Opisthokonta; Fungi; Dikarya; Ascomycota; saccharomyceta; Saccharomycotina; Saccharomycetes; Saccharomycetales; Saccharomycetaceae; Kluyveromyces
Kluyveromyces lactis 

')

```

```{r}
#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  # 
# Multiple Sequence Alignment
#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  # 

#############################################
## DECIPHER
#############################################

# I am using here DECIPHER, it outperforms MUSCLE at large datasets
# It has a pretty decent optimisation done with a single function
# We can also do synteny block analysis, which is awesome!

sequences <- "/home/mrinalmanu/Documents/3_sem_hws/phylo_3/data/SUP35_10seqs.fa"
dna <- readDNAStringSet(sequences)
dna # the unaligned sequences

```

```{r}
DNA <- AlignSeqs(dna) # align the sequences directly without translation

DNA <- AlignTranslation(dna) # align the translation then reverse translate
```

```{r}

BrowseSeqs(DNA, highlight=1) # view the alignment

```

```{r}
# form a chained guide tree
gT <- lapply(order(width(dna), decreasing=TRUE),
  function(x) {
  attr(x, "height") <- 0
  attr(x, "label") <- names(dna)[x]
  attr(x, "members") <- 1L
  attr(x, "leaf") <- TRUE
  x
  })

attr(gT, "height") <- 0.5
attr(gT, "members") <- length(dna)
class(gT) <- "dendrogram"

##### The aligment imporves based on the guide tree
# use the guide tree as input for alignment
dc_start_time <- Sys.time()

DNA <- AlignTranslation(dna,
        guideTree=gT,
        iterations=0,
        refinements=0)

dc_end_time <- Sys.time()

dc_ex_t <- dc_end_time - dc_start_time

# Visualization using the default theme named theme_dendro()
ggdendrogram(gT, rotate = TRUE, theme_dendro = FALSE)

```


```{r}

# let's optimize our alignment

BrowseSeqs(DNA, highlight=0)

DNA_adjusted <- AdjustAlignment(DNA)

# Tending to the problem of false homologies; one function can do this automatically!
DNA_staggered <- StaggerAlignment(DNA)

```

```{r}
#############################################
## Muscle
#############################################
ms_start_time <- Sys.time()

muscle_alignment <- muscle(dna)

ms_end_time <- Sys.time()

ms_ex_t <- ms_end_time - ms_start_time

msaPrettyPrint(muscle_alignment, y=c(164, 213), output="asis", showNames="none", showLogo="none", askForOverwrite=FALSE)
```

```{r}

muscle_alignment_with_penalty <- muscle(dna, gapopen = -10)

msaPrettyPrint(muscle_alignment_with_penalty, y=c(164, 213), output="asis", showNames="none", showLogo="none", askForOverwrite=FALSE)
```

```{r}
#############################################
## Clustal Omega
#############################################

strings_given <- dna
clusta_w <- msa(strings_given, "ClustalW")

msaPrettyPrint(clusta_w, y=c(164, 213), output="asis", showNames="none", showLogo="none", askForOverwrite=FALSE)
```
```{r}
cs_start_time <- Sys.time()

clusta_w_cool <- msa(strings_given, "ClustalW", gapOpening = -10)

cs_end_time <- Sys.time()

cs_ex_time <- cs_end_time - cs_start_time

msaPrettyPrint(clusta_w_cool, y=c(164, 213), output="asis", showNames="none", showLogo="none", askForOverwrite=FALSE)


```

```{r}
#############################################
## Clustal
#############################################
c_start_time <- Sys.time()

ALN <- msa(strings_given)

c_end_time <- Sys.time()

c_ex_time <- c_start_time - c_end_time

ALN

```
```{r}
print('time for DECIPHER:')
dc_ex_t
print('time for Muscle:')
ms_ex_t
print('time for Clustal_omega:')
cs_ex_time 
print('time for Clustal:')
c_ex_time

# With optimisation I find Decipher as the best. 
```

```{r}
#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  # 
# What is wrong with the alignment of SUP35_10seqs_strange_aln.fa and how to fix it?
#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  # 

new_sequences <- "/home/mrinalmanu/Documents/3_sem_hws/phylo_3/data/SUP35_10seqs_strange_aln.fa"
new_dna <- readDNAStringSet(new_sequences)
print('We can see that here one of the strings is in reverse. This is what is causing the strange behaviour of this sequence:')
new_dna[4]

print('All other sequences are like this:')
new_dna[1]
new_dna[2]

print("We should reverse this sequence, or use complement.")

```

```{r}
#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  # 
# Repeating Decipher for protein sequence
#  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  #  # 

AA <- AlignTranslation(dna, type="AAStringSet") # align the translation
BrowseSeqs(AA, highlight=1) # view the alignment

```

  